graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/TopSecondDegreeByCount.java

```diff
@@ -61,96 +43,109 @@
+
+    collectRightNodeInfo(request, visitedRightNodes);
+    recordIntermediateStats(request.getQueryNode(), visitedRightNodes);
+    filterNodeInfo(request, visitedRightNodes, nodeInfosAfterFiltering);
+    return generateRecommendationFromNodeInfo(request, nodeInfosAfterFiltering);
+  }
+
+  private void reset(Request request) {
+    request.resetFilters();
+    visitedRightNodes.clear();
+    nodeInfosAfterFiltering.clear();
+    seenEdgesPerNode.clear();
+    topSecondDegreeByCountStats.reset();
+  }
+
+  private void collectRightNodeInfo(Request request, Long2ObjectMap<NodeInfo> collectedNodeInfo) {
     for (Long2DoubleMap.Entry entry: request.getLeftSeedNodesWithWeight().long2DoubleEntrySet()) {
       long leftNode = entry.getLongKey();
       double weight = entry.getDoubleValue();
       int numEdgesPerNode = 0;
       NodeMetadataMultiSegmentIterator edgeIterator =
-        (NodeMetadataMultiSegmentIterator) leftIndexedBipartiteGraph.getLeftNodeEdges(leftNode);
+          (NodeMetadataMultiSegmentIterator) leftIndexedBipartiteGraph.getLeftNodeEdges(leftNode);
       seenEdgesPerNode.clear();
 
-      if (edgeIterator != null) {
-        // Sequentially iterating through the latest MAX_EDGES_PER_NODE edges per node
-        while (edgeIterator.hasNext() && numEdgesPerNode++ < MAX_EDGES_PER_NODE) {
-          long rightNode = edgeIterator.nextLong();
-          byte edgeType = edgeIterator.currentEdgeType();
-
-          if (seenEdgesPerNode.containsKey(rightNode)
-            && seenEdgesPerNode.get(rightNode) == edgeType) {
-            // no op
-          } else {
-            seenEdgesPerNode.put(rightNode, edgeType);
-
-            NodeInfo nodeInfo;
-            if (!visitedRightNodes.containsKey(rightNode)) {
-              int metadataSize = RecommendationType.METADATASIZE.getValue();
-
-              int[][] nodeMetadata = new int[metadataSize][];
-
-              for (int i = 0; i < metadataSize; i++) {
-                IntArrayIterator metadataIterator =
-                  (IntArrayIterator) edgeIterator.getRightNodeMetadata((byte) i);
-
-                if (metadataIterator.size() > 0) {
-                  int[] metadata = new int[metadataIterator.size()];
-                  int j = 0;
-                  while (metadataIterator.hasNext()) {
-                    metadata[j++] = metadataIterator.nextInt();
-                  }
-                  nodeMetadata[i] = metadata;
-                }
-              }
-
-              nodeInfo = new NodeInfo(
-                rightNode,
-                nodeMetadata,
-                0.0,
-                request.getMaxSocialProofTypeSize()
-              );
-              visitedRightNodes.put(rightNode, nodeInfo);
-            } else {
-              nodeInfo = visitedRightNodes.get(rightNode);
-            }
-
-            nodeInfo.addToWeight(weight);
-            nodeInfo.addToSocialProof(leftNode, edgeType, weight);
-          }
+      if (edgeIterator == null) {
```

Comment:
Personally, I favor guard conditions than nested if. Breaking the original format doesn't necessarily make the code harder to read but actually easier imo, because it conveys the same logic while avoiding a layer nesting. If this makes the code hard to read I will change it back, but I think in this case it is justified